// @nomiclabs/buidler/internal/buidler-evm/stack-traces/solidity-stack-trace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var StackTraceEntryType;
(function (StackTraceEntryType) {
    StackTraceEntryType[StackTraceEntryType["CALLSTACK_ENTRY"] = 0] = "CALLSTACK_ENTRY";
    StackTraceEntryType[StackTraceEntryType["UNRECOGNIZED_CREATE_CALLSTACK_ENTRY"] = 1] = "UNRECOGNIZED_CREATE_CALLSTACK_ENTRY";
    StackTraceEntryType[StackTraceEntryType["UNRECOGNIZED_CONTRACT_CALLSTACK_ENTRY"] = 2] = "UNRECOGNIZED_CONTRACT_CALLSTACK_ENTRY";
    StackTraceEntryType[StackTraceEntryType["PRECOMPILE_ERROR"] = 3] = "PRECOMPILE_ERROR";
    StackTraceEntryType[StackTraceEntryType["REVERT_ERROR"] = 4] = "REVERT_ERROR";
    StackTraceEntryType[StackTraceEntryType["FUNCTION_NOT_PAYABLE_ERROR"] = 5] = "FUNCTION_NOT_PAYABLE_ERROR";
    StackTraceEntryType[StackTraceEntryType["INVALID_PARAMS_ERROR"] = 6] = "INVALID_PARAMS_ERROR";
    StackTraceEntryType[StackTraceEntryType["FALLBACK_NOT_PAYABLE_ERROR"] = 7] = "FALLBACK_NOT_PAYABLE_ERROR";
    StackTraceEntryType[StackTraceEntryType["UNRECOGNIZED_FUNCTION_WITHOUT_FALLBACK_ERROR"] = 8] = "UNRECOGNIZED_FUNCTION_WITHOUT_FALLBACK_ERROR";
    StackTraceEntryType[StackTraceEntryType["RETURNDATA_SIZE_ERROR"] = 9] = "RETURNDATA_SIZE_ERROR";
    StackTraceEntryType[StackTraceEntryType["NONCONTRACT_ACCOUNT_CALLED_ERROR"] = 10] = "NONCONTRACT_ACCOUNT_CALLED_ERROR";
    StackTraceEntryType[StackTraceEntryType["CALL_FAILED_ERROR"] = 11] = "CALL_FAILED_ERROR";
    StackTraceEntryType[StackTraceEntryType["DIRECT_LIBRARY_CALL_ERROR"] = 12] = "DIRECT_LIBRARY_CALL_ERROR";
    StackTraceEntryType[StackTraceEntryType["UNRECOGNIZED_CREATE_ERROR"] = 13] = "UNRECOGNIZED_CREATE_ERROR";
    StackTraceEntryType[StackTraceEntryType["UNRECOGNIZED_CONTRACT_ERROR"] = 14] = "UNRECOGNIZED_CONTRACT_ERROR";
    StackTraceEntryType[StackTraceEntryType["OTHER_EXECUTION_ERROR"] = 15] = "OTHER_EXECUTION_ERROR";
})(StackTraceEntryType = exports.StackTraceEntryType || (exports.StackTraceEntryType = {}));
exports.FALLBACK_FUNCTION_NAME = "<fallback>";
exports.CONSTRUCTOR_FUNCTION_NAME = "constructor";
exports.UNRECOGNIZED_FUNCTION_NAME = "<unrecognized-selector>";
exports.UNKNOWN_FUNCTION_NAME = "<unknown>";
exports.PRECOMPILE_FUNCTION_NAME = "<precompile>";
exports.UNRECOGNIZED_CONTRACT_NAME = "<UnrecognizedContract>";
