// @studydefi/money-legos/compound/contracts.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var Comptroller_json_1 = __importDefault(require("./abi/Comptroller.json"));
var CompoundPriceOracle_json_1 = __importDefault(require("./abi/CompoundPriceOracle.json"));
var CToken_json_1 = __importDefault(require("./abi/CToken.json"));
var CEther_json_1 = __importDefault(require("./abi/CEther.json"));
var contracts = {
    cToken: { abi: CToken_json_1.default },
    comptroller: {
        address: "0x3d9819210a31b4961b30ef54be2aed79b9c9cd3b",
        abi: Comptroller_json_1.default,
    },
    priceOracle: {
        address: "0x1d8aedc9e924730dd3f9641cdb4d1b92b848b4bd",
        abi: CompoundPriceOracle_json_1.default,
    },
    cEther: {
        decimals: 8,
        address: "0x4Ddc2D193948926D02f9B1fE9e1daa0718270ED5",
        abi: CEther_json_1.default,
    },
    cDAI: {
        decimals: 8,
        address: "0x5d3a536E4D6DbD6114cc1Ead35777bAB948E3643",
        abi: CToken_json_1.default,
    },
    cSAI: {
        decimals: 8,
        address: "0xf5dce57282a584d2746faf1593d3121fcac444dc",
        abi: CToken_json_1.default,
    },
    cBAT: {
        decimals: 8,
        address: "0x6c8c6b02e7b2be14d4fa6022dfd6d75921d90e4e",
        abi: CToken_json_1.default,
    },
    cZRX: {
        decimals: 8,
        address: "0xb3319f5d18bc0d84dd1b4825dcde5d5f7266d407",
        abi: CToken_json_1.default,
    },
    cUSDC: {
        decimals: 8,
        address: "0x39aa39c021dfbae8fac545936693ac917d5e7563",
        abi: CToken_json_1.default,
    },
    cREP: {
        decimals: 8,
        address: "0x158079ee67fce2f58472a96584a73c7ab9ac95c1",
        abi: CToken_json_1.default,
    },
    cWBTC: {
        decimals: 8,
        address: "0xc11b1268c1a384e55c48c2391d8d480264a3a7f4",
        abi: CToken_json_1.default,
    },
};
exports.default = contracts;
